core/src/main/kotlin/com/github/quillraven/quillysadventure/screen/LoadingScreen.kt

```diff
@@ -119,12 +124,9 @@ class LoadingScreen(
     }
 
     override fun render(delta: Float) {
-        assets.update()
-        if (!loaded) {
-            loadingBar.scaleTo(assets.progress)
-        }
+        loadingBar.scaleTo(assets.progress.percent)
 
-        if (assets.progress >= 1f && !loaded) {
+        if (assets.progress.isFinished && !loaded) {
```

Comment:
Just one comment - `asset.progress.isFinished` will mostly work fine, but you might also try to use coroutines to await for the results instead. You won't need the `loaded` flag at all, nor will you have to check if the loading is finished on every render.

To do that:

* Save results from `loadAsync` to a collection (e.g. `MutableList<Deferred<*>>`) near lines 96-106.
* Use `joinAll()` to await for all the assets:

```
KtxAsync.launch {
   results.joinAll() // Suspends until all scheduled assets are loaded.
   // Now you can do the stuff you normally perform when loading is finished.
}
```

Benefits? You no longer have to maintain the `loaded` flag and worry that the initialization code gets executed twice. Also, since the assets are scheduled asynchronously, there's no way that `isFinished` reports as true even if the assets are not loaded (this should be extremely rare, but still).

Reply:
#41